x/liquidity/keeper/migration.go

```diff
@@ -1,22 +1,164 @@
 package keeper
 
 import (
+	"fmt"
+	"strings"
+
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
 
 	v043 "github.com/gravity-devs/liquidity/x/liquidity/legacy/v043"
+	liquiditytypes "github.com/gravity-devs/liquidity/x/liquidity/types"
 )
 
 // Migrator is a struct for handling in-place store migrations.
 type Migrator struct {
-	keeper Keeper
+	keeper        Keeper
+	bankKeeper    liquiditytypes.BankKeeper
+	accountKeeper liquiditytypes.AccountKeeper
 }
 
 // NewMigrator returns a new Migrator.
-func NewMigrator(keeper Keeper) Migrator {
-	return Migrator{keeper: keeper}
+func NewMigrator(keeper Keeper, bankKeeper liquiditytypes.BankKeeper, accountKeeper liquiditytypes.AccountKeeper) Migrator {
+	return Migrator{keeper: keeper, bankKeeper: bankKeeper, accountKeeper: accountKeeper}
 }
 
 // Migrate1to2 migrates from version 1 to 2.
 func (m Migrator) Migrate1to2(ctx sdk.Context) error {
 	return v043.MigrateStore(ctx, m.keeper.storeKey)
 }
+
+// Migrate2to3 migrates from version 2 to 3.
+func (m Migrator) Migrate2to3(ctx sdk.Context) error {
+	return SafeForceWithdrawal(ctx, m.keeper, m.bankKeeper, m.accountKeeper)
+}
+
+// SafeForceWithdrawal call ForceWithdrawal safely by recover, cached ctx
+func SafeForceWithdrawal(ctx sdk.Context, keeper Keeper, bankKeeper liquiditytypes.BankKeeper, accountKeeper liquiditytypes.AccountKeeper) error {
+	broken := false
+	defer func() {
+		if r := recover(); r != nil {
+			broken = true
+		}
+	}()
+
+	cachedCtx, writeCache := ctx.CacheContext()
+	err := ForceWithdrawal(cachedCtx, keeper, bankKeeper, accountKeeper)
+	if err == nil && !broken {
```

Comment:
non-blocking: the `broken` local variable can be safely removed since this condition and the rest of the function won't be executed if either `ForceWithdrawal()` or `ctx.CacheContext()` panics.